### PR TITLE
[cli][fs-detectors][frameworks] Add Storybook support

### PR DESCRIPTION
**Pull request by cb1kenobi**
Adds support for Storybook projects. Normally, the fs detector will see the project as a Next.js (or something else) and this can cause issues. For example, if the project has a `middleware.ts` file, a corresponding function will be generated and interfere with the Storybook site.

The general idea is a user would create two projects associated to the same repo. The first one would be for the main site. The second one is for the storybook site. In the second site's project settings, the user would manually set the framework to "Storybook" as well as override the build command (e.g. `npm run storybook-build`). The user can also manually set the framework in the project settings (or `.vercel/project.json`) to `"storybook"`.

The Storybook framework is configured to treat the project as a static build which will exclude middleware, functions, etc.

Ref: https://github.com/orgs/vercel/discussions/1321

### CODE REVIEW

File: packages/frameworks/src/frameworks.ts [hunk 1]
```diff
@@ -1932,6 +1932,33 @@ export const frameworks = [
...
+  {
+    name: 'Storybook',
+    slug: 'storybook',
+    logo: 'https://api-frameworks.vercel.sh/framework-logos/storybook.svg',
+    tagline: 'Frontend workshop for UI development',
+    description:
+      'Storybook is a frontend workshop for building UI components and pages in isolation.',
+    website: 'https://storybook.js.org/',
+    ignoreRuntimes: ['@vercel/next', '@vercel/node'],
+    useRuntime: { src: 'package.json', use: '@vercel/static-build' },
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]: You can remove `useRuntime` since that is implied

File: packages/frameworks/src/frameworks.ts [hunk 2]
```diff
@@ -1932,6 +1932,33 @@ export const frameworks = [
...
     ],
   },
+  {
+    name: 'Storybook',
+    slug: 'storybook',
+    logo: 'https://api-frameworks.vercel.sh/framework-logos/storybook.svg',
+    tagline: 'Frontend workshop for UI development',
+    description:
+      'Storybook is a frontend workshop for building UI components and pages in isolation.',
+    website: 'https://storybook.js.org/',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]:
```suggestion
    website: 'https://storybook.js.org',
```

File: packages/frameworks/src/frameworks.ts [hunk 3]
```diff
@@ -1932,6 +1932,33 @@ export const frameworks = [
...
+    ignoreRuntimes: ['@vercel/next', '@vercel/node'],
+    useRuntime: { src: 'package.json', use: '@vercel/static-build' },
+    settings: {
+      installCommand: {
+        placeholder: '`yarn install`, `pnpm install`, or `npm install`',
+      },
+      buildCommand: {
+        placeholder: '`npm run build-storybook`',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]: placeholder should have `or` (see others)

File: packages/frameworks/src/frameworks.ts [hunk 4]
```diff
@@ -1932,6 +1932,33 @@ export const frameworks = [
...
+        value: 'npm run build-storybook',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]: value should not use `npm run` (see others)

File: packages/frameworks/src/frameworks.ts [hunk 5]
```diff
@@ -1932,6 +1932,33 @@ export const frameworks = [
...
+      },
+      devCommand: {
+        placeholder: '`npm run storybook`',
+        value: `npm run storybook`,
+      },
+      outputDirectory: {
+        value: null,
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]: I believe the output dir is required

File: packages/fs-detectors/src/detect-builders.ts [hunk 6]
```diff
@@ -531,6 +517,20 @@ function detectFrontBuilder(
...
+  ) {
+    const deps: PackageJson['dependencies'] = {
+      ...pkg.dependencies,
+      ...pkg.devDependencies,
+    };
+
+    if (deps['next']) {
+      framework = 'nextjs';
+    }
+  }
```
Comment (styfle) [inline packages/fs-detectors/src/detect-builders.ts]: Why was this moved?

File: packages/frameworks/src/frameworks.ts (hunk 1 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts]: Done!

File: packages/frameworks/src/frameworks.ts (hunk 3 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts]: Done!

File: packages/frameworks/src/frameworks.ts (hunk 4 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts]: Done!

File: packages/frameworks/src/frameworks.ts (hunk 5 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts]: Done!

File: packages/fs-detectors/src/detect-builders.ts (hunk 6 above)
Comment (cb1kenobi) [reply packages/fs-detectors/src/detect-builders.ts]: Reverted the change, but I think the code may be flawed. `createdAt` defaults to `0`, so if `framework` is valid, but `createdAt` is `undefined`, it'll force the framework to be `nextjs`.

File: packages/frameworks/src/frameworks.ts (hunk 4 above)
Comment (styfle) [reply packages/frameworks/src/frameworks.ts]: @cb1kenobi I still see `npm run build-storybook`

It should be `storybook build`

Also now that I think about it, the placeholder would be `npm run build` since thats the default.

File: packages/frameworks/src/frameworks.ts [hunk 7]
```diff
@@ -1932,6 +1932,32 @@ export const frameworks = [
...
+      installCommand: {
+        placeholder: '`yarn install`, `pnpm install`, or `npm install`',
+      },
+      buildCommand: {
+        placeholder:
+          '`npm run build-storybook`, `yarn build-storybook`, or `pnpm build-storybook`',
+        value: 'npm run build-storybook',
+      },
+      devCommand: {
+        value: `start-storybook -p 6006`,
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]: This should use a PORT env var (see other frameworks)

Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts]: Done!

File: packages/frameworks/src/frameworks.ts (hunk 4 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts]: Whoops, done!

File: packages/frameworks/src/frameworks.ts [hunk 8]
```diff
@@ -1932,6 +1932,30 @@ export const frameworks = [
       },
     ],
   },
+  {
+    name: 'Storybook',
```
Comment (EndangeredMassa) [inline packages/frameworks/src/frameworks.ts:1936]: To be clear: This means that Storybook is never auto-detected, because there are no detectors. I think this is what we want for now, but it's worth calling out.

**Review by EndangeredMassa — DISMISSED** [review]

File: packages/frameworks/src/frameworks.ts (hunk 8 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts:1936]: Correct.

Interestingly enough, we used to have Storybook in the list of frameworks, but it was removed (https://github.com/vercel/vercel/pull/3864) because it didn't have a detector.

Comment (styfle) [reply packages/frameworks/src/frameworks.ts:1936]: @cb1kenobi Lets add a comment here to explain that is intentional so it doesn't get removed in the future

Comment (styfle) [reply packages/frameworks/src/frameworks.ts:1936]: I'm also curious - why not have any detectors?

File: packages/frameworks/src/frameworks.ts [hunk 9]
```diff
@@ -1932,6 +1932,30 @@ export const frameworks = [
...
+    slug: 'storybook',
+    logo: 'https://api-frameworks.vercel.sh/framework-logos/storybook.svg',
+    tagline: 'Frontend workshop for UI development',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts:1939]: Are you missing the `demo` prop?

File: packages/frameworks/src/frameworks.ts (hunk 8 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts:1936]: I added the detector. Originally I didn't think it mattered since according to https://storybook.js.org/docs/react/get-started/install:

> Storybook needs to be installed into a project that is already set up with a framework.

In theory, we will detect the site as Next.js or whatever before we would detect it as Storybook.

However, maybe the user is doing something special like the UI components are in an external package or something which would cause the project to be detected as Storybook.

File: packages/frameworks/src/frameworks.ts (hunk 9 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts:1939]: `demo` is not required and I didn't prepare a demo. Can that be added in a future PR?

File: packages/frameworks/src/frameworks.ts [hunk 10]
```diff
@@ -1932,6 +1932,37 @@ export const frameworks = [
...
+    description:
+      'Storybook is a frontend workshop for building UI components and pages in isolation.',
+    website: 'https://storybook.js.org',
+    ignoreRuntimes: ['@vercel/next', '@vercel/node'],
+    detectors: {
+      every: [
+        {
+          matchPackage: 'storybook/.+',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]:
```suggestion
          matchPackage: '@storybook/cli',
```

Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts]: This is not correct. The `storybook` package is a wrapper for `@storybook/cli`. Users don't install the `storybook` package as a dependency. Instead they run `pnpm dlx storybook@latest init` which added project specific packages like react vs preact vs vue.

Comment (styfle) [reply packages/frameworks/src/frameworks.ts]: @cb1kenobi So in that case, it should check for `@storybook/cli`, right?

Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts]: OK, not sure what's up with my example, but I don't have `storybook` in the `package.json`, but it's in `node_modules`. I created a new app and now `storybook` is in the `package.json`. Weirdness. PR updated to detect `storybook`. Good catch.

**Review by EndangeredMassa — DISMISSED** [review]

File: packages/frameworks/src/frameworks.ts [hunk 11]
```diff
@@ -1932,6 +1932,37 @@ export const frameworks = [
...
+    },
+    settings: {
+      installCommand: {
+        placeholder: '`yarn install`, `pnpm install`, or `npm install`',
+      },
+      buildCommand: {
+        value: 'storybook build',
+      },
+      devCommand: {
+        value: `start-storybook -p $PORT`,
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]:
```suggestion
        value: `storybook dev -p $PORT`,
```

Where did you get `start-storybook` from? 

The `examples/storybook/package.json` seems to indicate this is `storybook dev`

File: packages/frameworks/src/frameworks.ts [hunk 12]
```diff
@@ -1932,6 +1932,37 @@ export const frameworks = [
...
+      },
+      outputDirectory: {
+        value: 'storybook-static',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts:1962]: Is the output dir configurable? We might need to read the config file to handle all cases

File: packages/frameworks/src/frameworks.ts (hunk 11 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts]: Good question. Copy and paste error.

File: packages/frameworks/src/frameworks.ts (hunk 12 above)
Comment (cb1kenobi) [reply packages/frameworks/src/frameworks.ts:1962]: There is a `.storybook/main.ts` where you tell Storybook which framework the project is using, where to find your stories, etc. It doesn't appear that the output directory is one of the configurable settings.

The `storybook build` command supports a `--output-dir [path]` option. The default value is `"storybook-static"`: https://github.com/storybookjs/storybook/blob/e7817e82757618992f930f6157ddb68fc7541260/code/lib/cli/src/build.ts#L9.

**Review by EndangeredMassa — APPROVED** [review]

**Comment by bencao** [comment]
Got the error "The provided framework must match one of the supported options." when I was trying to save a project's framework setting to "Storybook".

**Comment by cb1kenobi** [comment]
~~@bencao try `"storybook"` all lowercase.~~

**Comment by ovflowd** [comment]
Hey there 👋 is this released by now?

**Comment by ovflowd** [comment]
Oddly getting this here 👀 should I open an issue? Payload seems to be with lowercase "s"

image

image

**Comment by cb1kenobi** [comment]
@bencao, @ovflowd I just checked and the Storybook support added via this pull request has not released yet. I'm not quite sure why it's showing up in the dropdown. In any case, this will ship in the next CLI release.

**Comment by ovflowd** [comment]
Thanks @cb1kenobi any eta for that release? (When it usually happens 👀)
